nengo/connection.py

```diff
@@ -117,25 +117,70 @@ def validate(self, conn, distorarray):
 
+    def __set__(self, conn, function):
+        if function is None:
+            function_info = self.Info(function=None, size=None)
+        elif is_array_like(function):
+            array = np.array(function, copy=False, dtype=np.float64)
+            self.validate_array(conn, array)
+            function_info = self.Info(function=array, size=array.shape[1])
+        elif callable(function):
+            function_info = self.Info(function=function,
+                                      size=self.determine_size(conn, function))
+            self.validate_callable(conn, function_info)
+        else:
+            raise ValidationError("Invalid connection function type %r "
+                                  "(must be callable or array-like)"
+                                  % type(function).__name__,
+                                  attr=self.name, obj=conn)
+
+        self.validate(conn, function_info)
+        self.data[conn] = function_info
+
     def function_args(self, conn, function):
         x = (conn.eval_points[0] if is_iterable(conn.eval_points)
              else np.zeros(conn.size_in))
         return (x,)
 
-    def validate(self, conn, function_info):
-        super(ConnectionFunctionParam, self).validate(conn, function_info)
-        fn_ok = (Node, Ensemble)
-        function, size = function_info
+    def validate_array(self, conn, ndarray):
+        if not isinstance(conn.eval_points, np.ndarray):
+            raise ValidationError(
+                "Connection evaluation points must be explicitly set to be "
+                "an array, to set function points",
+                attr=self.name, obj=conn)
+
+        if ndarray.ndim != 2:
+            raise ValidationError("array must be 2D (got %dD)" % ndarray.ndim,
+                                  attr=self.name, obj=conn)
 
-        if function is not None and not isinstance(conn.pre_obj, fn_ok):
+        if ndarray.shape[0] != conn.eval_points.shape[0]:
             raise ValidationError(
-                "function can only be set for connections from an Ensemble or "
-                "Node (got type %r)" % type(conn.pre_obj).__name__,
+                "Number of evaluation points must match number "
+                "of function points (%d != %d)"
+                % (ndarray.shape[0], conn.eval_points.shape[0]),
                 attr=self.name, obj=conn)
 
+    def validate_callable(self, conn, function_info):
+        super(ConnectionFunctionParam, self).validate(conn, function_info)
+
+    def validate(self, conn, function_info):
+        function, size = function_info
         type_pre = type(conn.pre_obj).__name__
+
+        if function is not None:
+            if not isinstance(conn.pre_obj, (Node, Ensemble)):
+                raise ValidationError(
+                    "function can only be set for connections from an Ensemble"
+                    " or Node (got type %r)" % type_pre,
+                    attr=self.name, obj=conn)
```

Comment:
`Connection`, `Connection.learning_rule` and `Probe` can act as `post`, but I think we disallow them as `pre`.